baybe/searchspace/discrete.py

```diff
@@ -251,19 +255,33 @@ def from_product(
         empty_encoding: bool = False,
     ) -> SubspaceDiscrete:
         """See :class:`baybe.searchspace.core.SearchSpace`."""
-        # Set defaults
+        # Set defaults and order constraints
         constraints = constraints or []
+        constraints = sorted(
+            constraints,
+            key=lambda x: DISCRETE_CONSTRAINTS_FILTERING_ORDER.index(x.__class__),
+        )
 
-        # Create a dataframe representing the experimental search space
-        exp_rep = parameter_cartesian_prod_to_df(parameters)
-
-        # Remove entries that violate parameter constraints
-        _apply_constraint_filter(exp_rep, constraints)
+        if POLARS_INSTALLED and not os.environ.get("BAYBE_DEACTIVATE_POLARS"):
+            # Apply polars product and filtering
+            lazy_df = parameter_cartesian_prod_polars(parameters)
+            mask: list[bool] = []
+            lazy_df = _apply_constraint_filter_polars(lazy_df, constraints, mask)
```

Comment:
the thing with the `mask` here is a slightly ugly C++ like workaround
it arises from the way we handle unimplemented polars cosntraints, but the current version was suggeted by @AdrianSosic  I didnt want to change it again and instead opted for this here, is there a better variant?

Reply:
Can you briefly explain why this is necessary *now*? Was this just found/observed during this PR or was this issue not existing previously?

Reply:
this workaround become necessary if you want to avoid that the second filtering call (pandas) repeats the constraints already processed during the first call. Previously this was simply ignores, ie cosntraints were doubly processed, hence no workaround needed

it arises from the fact that we let the base class throw a `NotImplementedError` if the polars version is not present. While this makes the treatment easy it doesnt really give a nice way of splitting up the constraints lists into pandas/polars parts

Im ope t a nicer suggestion for achieving this

Reply:
@AVHopp please resolve if this explains your question

Reply:
Yes, I also have to say that the workaround is quite ugly, tbh 😄 But could it be that you make your life much more complicated then necessary? The workaround seems only necessary because – for some reason I don't understand – you desperately seem to want apply the constraint through `_apply_constraint_filter_polars` as a batch, but whose only purpose is really to apply the constraints one after another and updating the mask, thereby causing the overhead!

Why not simply get rid of this method altogether which would
* require significantly less code
* removes the necessary for the mask, since you can simply apply the constraints one after another in the calling method, where you already have them available as collection anyway?

Reply:
I do not desperately want to use the utility and thought about getting rid of it, but the utility is for instance used in other places like tests

Reply:
Ok, if you want to keep the ability of applying the constraints as a batch, then I'd still suggest let's clean up a few things and avoid this C-like workaround and make it a bit more pythonic.

* First issue: the constraints in the utility are types as a `Collection`, but clearly the order matters here, so let's turn this into a `Sequence`
* Then, instead of passing the optional list that we modify in-place and of which you have no real control inside the function, I suggest to simply have a second output in the form of a tuple of bools that is created inside the function, and takes over the same role. That way, there is no coupling / side-effect beyond the scope of the function, which I think is a bit nicer and fulfills the exact same purpose. And having two outputs here is acceptable I'd say, because as a caller, you really *expect* two pieces of information.

What do you say?

Reply:
Would also make the code much shorter and requires less explanations ...

Reply:
surprised about this suggestion, I did not go for that because it makes the output of `_apply_constraints_*` functions inconsistent

Reply:
True, but is that an issue here? Both are private, and they *do* differ in the fact that one can always be applied while the other cannot. So they cannot be swapped against each other without additional machinery around – hence, even though the name suggest they do the same, this is not quite the case...

Reply:
And unrelated: not sure if you prefer it, but there is in fact the `itertools.compress` function which you could use on the calling side. Up to you

Reply:
no, what I meant is, you could simply pass that returned tuple to compress, instead of requiring zip-list-comp, since it's made exactly for that. but really a matter of taste...